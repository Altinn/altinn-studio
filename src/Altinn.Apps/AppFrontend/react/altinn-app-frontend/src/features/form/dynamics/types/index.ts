export interface IConditionalRenderingRules {
  [id: string]: IConditionalRenderingRule;
}

export interface IConditionalRenderingRule {
  selectedFunction: string;
  selectedAction: string;
  selectedFields: ISelectedFields;
  inputParams: IParameters;
  repeatingGroup?: IConditionalRenderingRepeatingGroup;
}

export interface IConditionalRenderingRepeatingGroup {
  groupId: string;
}

export interface IParameters {
  [id: string]: string;
}

export interface ISelectedFields {
  [id: string]: string;
}
