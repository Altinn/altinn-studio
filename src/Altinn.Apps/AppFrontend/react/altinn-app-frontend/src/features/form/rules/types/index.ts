export interface IRuleModelFieldElement {
  type: string;
  name: string;
  inputs: any;
}
