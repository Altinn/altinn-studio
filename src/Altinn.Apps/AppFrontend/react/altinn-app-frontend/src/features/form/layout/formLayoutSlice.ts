/* eslint-disable max-len */
import { createAction, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { ILayoutSets, IUiConfig } from 'src/types';
import { ILayouts } from './index';
import * as LayoutTypes from './formLayoutTypes';

export interface ILayoutState {
  layouts: ILayouts;
  error: Error;
  uiConfig: IUiConfig;
  layoutsets: ILayoutSets;
}

const initialState: ILayoutState = {
  layouts: null,
  error: null,
  uiConfig: {
    focus: null,
    hiddenFields: [],
    autoSave: null,
    repeatingGroups: {},
    currentView: 'FormLayout',
    navigationConfig: {},
    layoutOrder: null,
    pageTriggers: [],
  },
  layoutsets: null,
};

const moduleName = 'formLayout';

const formLayoutSlice = createSlice({
  name: moduleName,
  initialState,
  reducers: {
    fetchLayoutFulfilled: (state, action: PayloadAction<LayoutTypes.IFetchLayoutFulfilled>) => {
      const { layouts, navigationConfig } = action.payload;
      state.layouts = layouts;
      state.uiConfig.navigationConfig = navigationConfig;
      state.uiConfig.layoutOrder = Object.keys(layouts);
      state.error = null;
    },
    fetchLayoutRejected: (state, action: PayloadAction<LayoutTypes.IFormLayoutActionRejected>) => {
      const { error } = action.payload;
      state.error = error;
    },
    fetchLayoutSetsFulfilled: (state, action: PayloadAction<LayoutTypes.IFetchLayoutSetsFulfilled>) => {
      const { layoutSets } = action.payload;
      state.layoutsets = layoutSets;
    },
    fetchLayoutSetsRejected: (state, action: PayloadAction<LayoutTypes.IFormLayoutActionRejected>) => {
      const { error } = action.payload;
      state.error = error;
    },
    fetchLayoutSettingsFulfilled: (state, action: PayloadAction<LayoutTypes.IFetchLayoutSettingsFulfilled>) => {
      const { settings } = action.payload;
      if (settings && settings.pages) {
        state.uiConfig.hideCloseButton = settings?.pages?.hideCloseButton;
        state.uiConfig.pageTriggers = settings.pages.triggers;
        if (settings.pages.order) {
          state.uiConfig.layoutOrder = settings.pages.order;
          if (state.uiConfig.currentViewCacheKey) {
            state.uiConfig.currentView = localStorage.getItem(state.uiConfig.currentViewCacheKey)
              || settings.pages.order[0];
          } else {
            state.uiConfig.currentView = settings.pages.order[0];
          }
        }
      }
    },
    fetchLayoutSettingsRejected: (state, action: PayloadAction<LayoutTypes.IFormLayoutActionRejected>) => {
      const { error } = action.payload;
      state.error = error;
    },
    setCurrentViewCacheKey: (state, action: PayloadAction<LayoutTypes.ISetCurrentViewCacheKey>) => {
      const { key } = action.payload;
      state.uiConfig.currentViewCacheKey = key;
    },
    updateAutoSave: (state, action: PayloadAction<LayoutTypes.IUpdateAutoSave>) => {
      const { autoSave } = action.payload;
      state.uiConfig.autoSave = autoSave;
    },
    updateCurrentViewFulfilled: (state, action: PayloadAction<LayoutTypes.IUpdateCurrentViewFulfilled>) => {
      const { newView, returnToView } = action.payload;
      state.uiConfig.currentView = newView;
      state.uiConfig.returnToView = returnToView;
    },
    updateCurrentViewRejected: (state, action: PayloadAction<LayoutTypes.IFormLayoutActionRejected>) => {
      const { error } = action.payload;
      state.error = error;
    },
    updateFocusFulfilled: (state, action: PayloadAction<LayoutTypes.IUpdateFocusFulfilled>) => {
      const { focusComponentId } = action.payload;
      state.uiConfig.focus = focusComponentId;
    },
    updateFocusRejected: (state, action: PayloadAction<LayoutTypes.IFormLayoutActionRejected>) => {
      const { error } = action.payload;
      state.error = error;
    },
    updateHiddenComponents: (state, action: PayloadAction<LayoutTypes.IUpdateHiddenComponents>) => {
      const { componentsToHide } = action.payload;
      state.uiConfig.hiddenFields = componentsToHide;
    },
    updateRepeatingGroupsFulfilled: (state, action: PayloadAction<LayoutTypes.IUpdateRepeatingGroupsFulfilled>) => {
      const { repeatingGroups } = action.payload;
      state.uiConfig.repeatingGroups = repeatingGroups;
    },
    updateRepeatingGroupsRejected: (state, action: PayloadAction<LayoutTypes.IFormLayoutActionRejected>) => {
      const { error } = action.payload;
      state.error = error;
    },
    updateRepeatingGroupsEditIndexFulfilled: (state, action: PayloadAction<LayoutTypes.IUpdateRepeatingGroupsEditIndexFulfilled>) => {
      const { group, index } = action.payload;
      state.uiConfig.repeatingGroups[group].editIndex = index;
    },
    updateRepeatingGroupsEditIndexRejected: (state, action: PayloadAction<LayoutTypes.IFormLayoutActionRejected>) => {
      const { error } = action.payload;
      state.error = error;
    },
    calculatePageOrderAndMoveToNextPageFulfilled: (state, action: PayloadAction<LayoutTypes.ICalculatePageOrderAndMoveToNextPageFulfilled>) => {
      const { order } = action.payload;
      state.uiConfig.layoutOrder = order;
    },
    calculatePageOrderAndMoveToNextPageRejected: (state, action: PayloadAction<LayoutTypes.IFormLayoutActionRejected>) => {
      const { error } = action.payload;
      state.error = error;
    },
  },
});

const actions = {
  calculatePageOrderAndMoveToNextPage: createAction<LayoutTypes.ICalculatePageOrderAndMoveToNextPage>(`${moduleName}/calculatePageOrderAndMoveToNextPage`),
  fetchLayout: createAction(`${moduleName}/fetchLayout`),
  fetchLayoutSets: createAction(`${moduleName}/fetchLayoutSets`),
  fetchLayoutSettings: createAction(`${moduleName}/fetchLayoutSettings`),
  updateCurrentView: createAction<LayoutTypes.IUpdateCurrentView>(`${moduleName}/updateCurrentView`),
  updateFocus: createAction<LayoutTypes.IUpdateFocus>(`${moduleName}/updateFocus`),
  updateRepeatingGroups: createAction<LayoutTypes.IUpdateRepeatingGroups>(`${moduleName}/updateRepeatingGroups`),
  updateRepeatingGroupsEditIndex: createAction<LayoutTypes.IUpdateRepeatingGroupsEditIndex>(`${moduleName}/updateRepeatingGroupsEditIndex`),
  initRepeatingGroups: createAction(`${moduleName}/initRepeatingGroups`),
};

export const FormLayoutActions = {
  ...actions,
  ...formLayoutSlice.actions,
};

export default formLayoutSlice.reducer;
