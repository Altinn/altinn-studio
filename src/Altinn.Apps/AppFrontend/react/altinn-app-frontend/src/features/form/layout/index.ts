import { IOption, Triggers } from '../../../types';

export interface ILayouts {
  [id: string]: ILayout;
}

export interface ILayoutEntry {
  id: string;
  type?: string;
}

export interface ILayoutGroup extends ILayoutEntry {
  children: string[];
  dataModelBindings?: IDataModelBindings;
  maxCount: number;
  textResourceBindings?: ITextResourceBindings;
  tableHeaders?: string[];
}

export interface ILayoutComponent extends ILayoutEntry {
  type: string;
  dataModelBindings: IDataModelBindings;
  isValid?: boolean;
  readOnly: boolean;
  disabled?: boolean;
  required: boolean;
  textResourceBindings: ITextResourceBindings;
  triggers?: Triggers[];
}
export interface IDataModelBindings {
  [id: string]: string;
}

export interface ITextResourceBindings {
  [id: string]: string;
}

export type ILayout = Array<ILayoutComponent | ILayoutGroup>;

export interface ISelectionComponentProps extends ILayoutComponent {
  options?: IOption[];
  optionsId?: string;
}

export interface IDatepickerProps extends ILayoutComponent { }

export interface IFileuploadProps extends ILayoutComponent {
  maxNumberOfAttachments: number;
  maxFileSizeInMB: number;
  displayMode: any;
  hasCustomFileEndings: boolean;
  validFileEndings: any;
}

declare enum HeaderSize {
  S,
  M,
  L,
}

export interface IHeaderProps extends ILayoutComponent {
  size: HeaderSize;
}

declare enum InputFieldType {
  text,
  email,
  password,
}

export interface IInputProps extends ILayoutComponent {
  inputType: InputFieldType;
}

export interface INavigationButtonProps extends ILayoutComponent {
  next?: string;
  previous?: string;
}

export interface IParagraphProps extends ILayoutComponent { }

export interface IRadioButtonsProps extends ILayoutComponent {
  options: IOption[];
  preselectedOptionIndex: number;
}

export interface ITextAreaProps extends ILayoutComponent { }

export interface IAdressComponent extends ILayoutComponent {
  addressTextResourceBinding: string;
  areaCodeTextResourceBinding: string;
  coTextResourceBinding: string;
  simpleDisplayMode: boolean;
}
