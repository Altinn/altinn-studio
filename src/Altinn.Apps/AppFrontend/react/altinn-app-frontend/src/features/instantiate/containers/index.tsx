import { createStyles, withStyles, WithStyles } from '@material-ui/core/styles';
import { AxiosError } from 'axios';
import * as React from 'react';
import { useSelector } from 'react-redux';
import { AltinnAppHeader, AltinnModal, AltinnContentLoader, AltinnContentIconFormData } from 'altinn-shared/components';
import { Redirect } from 'react-router-dom';
import { AltinnAppTheme } from 'altinn-shared/theme';
import { IParty } from 'altinn-shared/types';
import { checkIfAxiosError } from 'altinn-shared/utils';
import { getTextFromAppOrDefault } from 'src/utils/textResource';
import { IAltinnWindow, IRuntimeState } from '../../../types';
import { changeBodyBackground } from '../../../utils/bodyStyling';
import { HttpStatusCodes } from '../../../utils/networking';
import { post } from '../../../utils/networking';
import InstantiationActions from '../instantiation/actions';
import MissingRolesError from './MissingRolesError';
import NoValidPartiesError from './NoValidPartiesError';
import UnknownError from './UnknownError';
import InstantiateValidationError from './InstantiateValidationError';

const styles = () => createStyles({
  modal: {
    boxShadow: null,
    MozBoxShadow: null,
    WebkitBoxShadow: null,
  },
});

const titleKey = 'instantiate.starting';

export interface IPartyValidation {
  valid: boolean;
  message: string;
  validParties: IParty[];
}

export interface IServiceInfoProps extends WithStyles<typeof styles> {
  // intentionally left empty
}

function InstantiateContainer(props: IServiceInfoProps) {
  changeBodyBackground(AltinnAppTheme.altinnPalette.primary.blue);
  const { org, app } = window as Window as IAltinnWindow;

  const [partyValidation, setPartyValidation] = React.useState(null);
  const [instantiating, setInstantiating] = React.useState(false);

  const instantiation = useSelector((state: IRuntimeState) => state.instantiation);
  const profile = useSelector((state: IRuntimeState) => state.profile.profile);
  const selectedParty = useSelector((state: IRuntimeState) => state.party.selectedParty);
  const titleText: any = useSelector((state: IRuntimeState) => {
    const text = getTextFromAppOrDefault(titleKey, state.textResources.resources, state.language.language, [], true);
    return text === titleKey ? '' : text;
  });

  const createNewInstance = () => {
    if (!selectedParty) {
      return;
    }
    setInstantiating(true);
    InstantiationActions.instantiate(org, app);
  };

  const validatatePartySelection = async () => {
    if (!selectedParty) {
      return;
    }
    try {
      const { data } = await post(
        `${window.location.origin}/${org}/${app}/api/v1/parties/` +
        `validateInstantiation?partyId=${selectedParty.partyId}`,
      );
      setPartyValidation(data);
    } catch (err) {
      console.error(err);
      throw new Error('Server did not respond with party validation');
    }
  };

  // eslint-disable-next-line no-undef
  const renderModalAndLoader = (): JSX.Element => {
    const { classes } = props;
    return (
      <>
        <AltinnModal
          classes={classes}
          isOpen={true}
          onClose={null}
          hideBackdrop={true}
          hideCloseIcon={true}
          headerText={titleText}
        >
          <AltinnContentLoader width='100%' height='400'>
            <AltinnContentIconFormData/>
          </AltinnContentLoader>
        </AltinnModal>
      </>
    );
  };

  React.useEffect(() => {
    validatatePartySelection();
  }, [selectedParty]);

  React.useEffect(() => {
    if (partyValidation !== null) {
      // validations
    }
  }, [partyValidation]);

  React.useEffect(() => {
    if (
      partyValidation !== null &&
      partyValidation.valid &&
      !instantiating &&
      !instantiation.instanceId
    ) {
      createNewInstance();
    }
  }, [partyValidation, instantiating, selectedParty]);

  if (instantiation.error !== null && checkIfAxiosError(instantiation.error)) {
    const axiosError = instantiation.error as AxiosError;
    const message = axiosError.response.data?.message;
    if (axiosError.response.status === HttpStatusCodes.Forbidden) {
      if (message) {
        return <InstantiateValidationError message={message} />;
      }
      return (
        <MissingRolesError />
      );
    }

    return (
      <UnknownError />
    );
  }

  if (partyValidation !== null && !partyValidation.valid) {
    if (partyValidation.validParties !== null &&
      partyValidation.validParties.length === 0) {
      return (
        <NoValidPartiesError />
      );
    }
    return (
      <Redirect to={`/partyselection/${HttpStatusCodes.Forbidden}`}/>
    );
  }

  if (instantiation.instanceId !== null) {
    return (
      <Redirect to={`/instance/${instantiation.instanceId}`} />
    );
  }

  return (
    <>
      <AltinnAppHeader
        logoColor={AltinnAppTheme.altinnPalette.primary.blueDarker}
        headerBackgroundColor={AltinnAppTheme.altinnPalette.primary.blue}
        party={selectedParty}
        userParty={profile ? profile.party : {} as IParty}
      />
      {renderModalAndLoader()}
    </>
  );
}

export default withStyles(styles)(InstantiateContainer);
