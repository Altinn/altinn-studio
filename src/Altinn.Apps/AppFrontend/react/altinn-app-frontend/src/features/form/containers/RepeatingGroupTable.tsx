/* eslint-disable no-undef */
/* eslint-disable react/no-array-index-key */
/* eslint-disable max-len */
import React from 'react';
import { Grid, makeStyles, createMuiTheme, TableContainer, Table, TableHead, TableRow, TableBody, TableCell, IconButton, useMediaQuery } from '@material-ui/core';
import altinnAppTheme from 'altinn-shared/theme/altinnAppTheme';
import { getLanguageFromKey } from 'altinn-shared/utils';
import { componentHasValidations, repeatingGroupHasValidations } from 'src/utils/validation';
import { createRepeatingGroupComponents } from 'src/utils/formLayout';
import { getFormDataForComponentInRepeatingGroup, getTextResource } from 'src/utils/formComponentUtils';
import { ILayout, ILayoutComponent, ILayoutGroup } from '../layout';
import { setupGroupComponents } from '../../../utils/layout';
import { ITextResource, IRepeatingGroups, IValidations, IOptions } from '../../../types';

export interface IRepeatingGroupTableProps {
  id: string;
  container: ILayoutGroup;
  components: (ILayoutComponent | ILayoutGroup)[]
  repeatingGroupIndex: number;
  repeatingGroups: IRepeatingGroups;
  hiddenFields: string[];
  formData: any;
  options: IOptions;
  textResources: ITextResource[];
  language: any;
  currentView: string;
  layout: ILayout;
  validations: IValidations;
  editIndex: number;
  setEditIndex: (index: number) => void;
  filteredIndexes?: number[];
}

const theme = createMuiTheme(altinnAppTheme);

const useStyles = makeStyles({
  table: {
    tableLayout: 'fixed',
    marginBottom: '12px',
    wordBreak: 'break-word',
  },
  tableHeader: {
    borderBottom: `2px solid ${theme.altinnPalette.primary.blueMedium}`,
    '& th': {
      fontSize: '1.4rem',
      padding: '0px',
      paddingLeft: '6px',
      '& p': {
        fontWeight: 500,
        fontSize: '1.4rem',
        padding: '0px',
        paddingLeft: '6px',
      },
    },
  },
  tableBody: {
    '& td': {
      borderBottom: `2px dotted ${theme.altinnPalette.primary.blueMedium}`,
      padding: '0px',
      paddingLeft: '6px',
      fontSize: '1.4rem',
      whiteSpace: 'nowrap',
      textOverflow: 'ellipsis',
      overflow: 'hidden',
    },
  },
  tableRowError: {
    backgroundColor: '#F9CAD3;',
  },
  errorIcon: {
    fontSize: '2em',
    minWidth: '0px',
    minHeight: '0px',
    width: 'auto',
  },
  editIcon: {
    paddingLeft: '6px',
  },
  mobileGrid: {
    borderBottom: `2px dotted ${theme.altinnPalette.primary.blueMedium}`,
    paddingLeft: '0.6rem',
    paddingRight: '0.6rem',
  },
  mobileContainer: {
    borderTop: `2px solid ${theme.altinnPalette.primary.blueMedium}`,
    marginBottom: '1.2rem',
  },
  mobileText: {
    fontWeight: 500,
    float: 'left',
    maxWidth: '50%',
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
    '& p': {
      fontWeight: 500,
    },
  },
  mobileValueText: {
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
    maxWidth: '50%',
    minWidth: '50%',
  },
  textContainer: {
    width: '100%',
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
  },
});

export function RepeatingGroupTable({
  id,
  container,
  components,
  repeatingGroupIndex,
  editIndex,
  formData,
  options,
  textResources,
  currentView,
  hiddenFields,
  language,
  layout,
  repeatingGroups,
  validations,
  setEditIndex,
  filteredIndexes,
}: IRepeatingGroupTableProps): JSX.Element {
  const classes = useStyles();
  const renderComponents: ILayoutComponent[] = JSON.parse(JSON.stringify(components));
  const tableHeaderComponents = container.tableHeaders
    || components.map((c) => (c as any).baseComponentId || c.id) || [];
  const mobileView = useMediaQuery('(max-width:992px)'); // breakpoint on altinn-modal
  const componentTitles: string[] = [];
  renderComponents.forEach((component: ILayoutComponent) => {
    const childId = (component as any).baseComponentId || component.id;
    if (tableHeaderComponents.includes(childId)) {
      componentTitles.push(component.textResourceBindings?.title || '');
    }
  });
  const repeatingGroupDeepCopyComponents = createRepeatingGroupComponents(
    container,
    renderComponents,
    repeatingGroupIndex,
    textResources,
    hiddenFields,
  );

  const getFormDataForComponent = (component: ILayoutComponent | ILayoutGroup, index: number): string => {
    return getFormDataForComponentInRepeatingGroup(formData, component, index, container.dataModelBindings.group, textResources, options);
  };

  const onClickEdit = (groupIndex: number) => {
    if (groupIndex === editIndex) {
      setEditIndex(-1);
    } else {
      setEditIndex(groupIndex);
    }
  };

  const childElementHasErrors = (element: ILayoutGroup | ILayoutComponent, index: number) => {
    if (element.type === 'Group') {
      return childGroupHasErrors(element as ILayoutGroup, index);
    }
    return componentHasValidations(validations, currentView, `${element.id}`);
  };

  const childGroupHasErrors = (childGroup: ILayoutGroup, index: number) => {
    const childGroupCount = repeatingGroups[childGroup.id]?.count;
    const childGroupComponents = layout.filter((childElement) => childGroup.children?.indexOf(childElement.id) > -1);
    const childRenderComponents = setupGroupComponents(childGroupComponents, childGroup.dataModelBindings?.group, index);
    const deepCopyComponents = createRepeatingGroupComponents(
      childGroup,
      childRenderComponents,
      childGroupCount,
      textResources,
      hiddenFields,
    );
    return repeatingGroupHasValidations(childGroup, deepCopyComponents, validations, currentView, repeatingGroups, layout, hiddenFields);
  };

  return (
    <Grid
      container={true}
      item={true}
      data-testid={`group-${id}`}
      id={`group-${id}`}
    >
      {!mobileView &&
      <TableContainer component={Grid}>
        <Table className={classes.table}>
          <TableHead className={classes.tableHeader}>
            <TableRow>
              {componentTitles.map((title: string) => (
                <TableCell align='left' key={title}>
                  {getTextResource(title, textResources)}
                </TableCell>
              ))}
              <TableCell/>
            </TableRow>
          </TableHead>
          <TableBody className={classes.tableBody}>
            {(repeatingGroupIndex >= 0) && [...Array(repeatingGroupIndex + 1)].map((_x: any, index: number) => {
              const rowHasErrors = repeatingGroupDeepCopyComponents[index].some((component: ILayoutComponent | ILayoutGroup) => {
                return childElementHasErrors(component, index);
              });

              // Check if filter is applied and includes specified index.
              if (filteredIndexes && !filteredIndexes.includes(index)) {
                return null;
              }

              return (
                <TableRow className={rowHasErrors ? classes.tableRowError : ''} key={index}>
                  {components.map((component: ILayoutComponent) => {
                    const childId = (component as any).baseComponentId || component.id;
                    if (!tableHeaderComponents.includes(childId)) {
                      return null;
                    }
                    return (
                      <TableCell key={`${component.id} ${index}`}>
                        {getFormDataForComponent(component, index)}
                      </TableCell>
                    );
                  })}
                  <TableCell align='right' key={`delete-${index}`}>
                    <IconButton style={{ color: 'black' }} onClick={() => onClickEdit(index)}>
                      {rowHasErrors ?
                        getLanguageFromKey('general.edit_alt_error', language) :
                        getLanguageFromKey('general.edit_alt', language)}
                      <i className={rowHasErrors ?
                        `ai ai-circle-exclamation a-icon ${classes.errorIcon} ${classes.editIcon}` :
                        `fa fa-editing-file ${classes.editIcon}`}
                      />
                    </IconButton>
                  </TableCell>
                </TableRow>);
            })}
          </TableBody>
        </Table>
      </TableContainer>}
      {mobileView &&
      <Grid
        container={true}
        item={true}
        direction='column'
        className={classes.mobileContainer}
      >
        {(repeatingGroupIndex >= 0) && [...Array(repeatingGroupIndex + 1)].map((_x: any, index: number) => {
          const rowHasErrors = repeatingGroupDeepCopyComponents[index].some((component: ILayoutComponent | ILayoutGroup) => {
            return childElementHasErrors(component, index);
          });
          return (
            <Grid
              item={true} container={true}
              justify='flex-end' direction='row'
              className={`${classes.mobileGrid} ${rowHasErrors ? classes.tableRowError : ''}`}
            >
              <Grid item={true}>
                <IconButton
                  style={{
                    color: 'black', padding: '0px', paddingLeft: '6px',
                  }} onClick={() => onClickEdit(index)}
                >
                  {rowHasErrors ?
                    getLanguageFromKey('general.edit_alt_error', language) :
                    getLanguageFromKey('general.edit_alt', language)}
                  <i className={rowHasErrors ?
                    `ai ai-circle-exclamation ${classes.errorIcon}` :
                    `fa fa-editing-file ${classes.editIcon}`}
                  />
                </IconButton>
              </Grid>
              {components.map((component: ILayoutComponent) => {
                const childId = (component as any).baseComponentId || component.id;
                if (!tableHeaderComponents.includes(childId)) {
                  return null;
                }
                return (
                  <Grid item={true} className={rowHasErrors ? `${classes.tableRowError} ${classes.textContainer}` : classes.textContainer}>
                    <div className={classes.mobileText}>
                      {getTextResource(component?.textResourceBindings?.title || '', textResources)}
                    </div>
                    <div
                      className={classes.mobileValueText}
                    >
                      {`: ${getFormDataForComponent(component, index)}`}
                    </div>
                  </Grid>
                );
              })}
            </Grid>
          );
        })}
      </Grid>
      }
    </Grid>
  );
}
