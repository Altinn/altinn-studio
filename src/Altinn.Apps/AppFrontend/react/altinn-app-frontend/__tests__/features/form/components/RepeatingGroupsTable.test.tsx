/* tslint:disable:jsx-wrap-multiline */
import '@testing-library/jest-dom/extend-expect';
import 'jest';
import * as React from 'react';
import { render } from '@testing-library/react';
import { RepeatingGroupTable } from '../../../../src/features/form/containers/RepeatingGroupTable';
import { ILayoutComponent, ILayoutGroup, ISelectionComponentProps } from '../../../../src/features/form/layout';
import { IOption, ITextResource } from '../../../../src/types';
import { ILayoutState } from '../../../../src/features/form/layout/formLayoutSlice';

describe('components/base/InputComponent.tsx', () => {
  let mockContainer: ILayoutGroup;
  let mockLanguage: any;
  let mockTextResources: ITextResource[];
  let mockComponents: ILayoutComponent[];
  let mockLayout: ILayoutState;
  let mockCurrentView: string;
  let mockData: any;
  let mockOptions: IOption[];

  beforeEach(() => {
    mockOptions = [{ value: 'option.value', label: 'option.label' }];
    mockComponents = [
      {
        id: 'field1',
        type: 'Input',
        dataModelBindings: {
          simpleBinding: 'Group.prop1',
        },
        textResourceBindings: {
          title: 'Title1',
        },
        readOnly: false,
        required: false,
        disabled: false,
      },
      {
        id: 'field2',
        type: 'Input',
        dataModelBindings: {
          simpleBinding: 'Group.prop2',
        },
        textResourceBindings: {
          title: 'Title2',
        },
        readOnly: false,
        required: false,
        disabled: false,
      },
      {
        id: 'field3',
        type: 'Input',
        dataModelBindings: {
          simpleBinding: 'Group.prop3',
        },
        textResourceBindings: {
          title: 'Title3',
        },
        readOnly: false,
        required: false,
        disabled: false,
      },
      {
        id: 'field4',
        type: 'Checkboxes',
        dataModelBindings: {
          simpleBinding: 'some-group.checkboxBinding',
        },
        textResourceBindings: {
          title: 'Title4',
        },
        readOnly: false,
        required: false,
        disabled: false,
        options: mockOptions,
      } as ISelectionComponentProps,
    ];

    mockContainer = {
      type: 'Group',
      id: 'mock-container-id',
      children: [
        'field1',
        'field2',
        'field3',
        'field4',
      ],
      maxCount: 8,
      dataModelBindings: {
        group: 'some-group',
      },
    };

    mockLayout = {
      layouts: {
        FormLayout: [].concat(mockContainer).concat(mockComponents),
      },
      uiConfig: {
        hiddenFields: [],
        repeatingGroups: {
          'mock-container-id': {
            count: 3,
          },
        },
        autoSave: false,
        currentView: 'FormLayout',
        focus: undefined,
        layoutOrder: ['FormLayout'],
      },
      error: null,
      layoutsets: null,
    };

    mockData = {
      formData: {
        'some-group[1].checkboxBinding': 'option.value',
      },
    };

    mockTextResources = [
      { id: 'option.label', value: 'Value to be shown' },
    ];

    mockCurrentView = 'FormLayout';
  });

  test('feature/form/containers/RepeatingGroupTable -- should match snapshot', () => {
    const { asFragment } = render(
      <RepeatingGroupTable
        container={mockContainer}
        language={mockLanguage}
        textResources={mockTextResources}
        components={mockComponents}
        currentView={mockCurrentView}
        editIndex={-1}
        formData={mockData}
        hiddenFields={[]}
        id={mockContainer.id}
        layout={mockLayout.layouts[mockCurrentView]}
        options={{}}
        repeatingGroupIndex={3}
        repeatingGroups={mockLayout.uiConfig.repeatingGroups}
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        setEditIndex={(index: number) => {}}
        validations={{}}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
