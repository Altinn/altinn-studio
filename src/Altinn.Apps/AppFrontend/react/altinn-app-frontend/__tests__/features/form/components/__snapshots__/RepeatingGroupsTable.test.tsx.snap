// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/base/InputComponent.tsx feature/form/containers/RepeatingGroupTable -- should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item"
    data-testid="group-mock-container-id"
    id="group-mock-container-id"
  >
    <div
      class="MuiGrid-root MuiTableContainer-root"
    >
      <table
        class="MuiTable-root makeStyles-table-1"
      >
        <thead
          class="MuiTableHead-root makeStyles-tableHeader-2"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft"
              scope="col"
            >
              Title1
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft"
              scope="col"
            >
              Title2
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft"
              scope="col"
            >
              Title3
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft"
              scope="col"
            >
              Title4
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head"
              scope="col"
            />
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root makeStyles-tableBody-3"
        >
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                style="color: black;"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  general.edit_alt
                  <i
                    class="fa fa-editing-file makeStyles-editIcon-6"
                  />
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                style="color: black;"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  general.edit_alt
                  <i
                    class="fa fa-editing-file makeStyles-editIcon-6"
                  />
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                style="color: black;"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  general.edit_alt
                  <i
                    class="fa fa-editing-file makeStyles-editIcon-6"
                  />
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body"
            />
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                style="color: black;"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  general.edit_alt
                  <i
                    class="fa fa-editing-file makeStyles-editIcon-6"
                  />
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
